spring-cloud-commons/src/main/java/reactor/core/publisher/FluxFirstNonEmptyEmitting.java

```diff
@@ -0,0 +1,344 @@
+
+			for (int i = 0; i < n; i++) {
+				a[i] = new FirstNonEmptyEmittingSubscriber<>(actual, this, i);
+			}
+
+			actual.onSubscribe(this);
+
+			for (int i = 0; i < n; i++) {
+				if (cancelled || wip != Integer.MIN_VALUE) {
+					return;
+				}
+
+				Publisher<? extends T> p = sources[i];
+
+				if (p == null) {
+					if (WIP.compareAndSet(this, Integer.MIN_VALUE, -1)) {
+						actual.onError(new NullPointerException(
+								"The " + i + " th Publisher source is null"));
+					}
+					return;
+				}
+
+				p.subscribe(a[i]);
+			}
+
+		}
+
+		@Override
+		public void request(long n) {
+			if (Operators.validate(n)) {
+				int w = wip;
+				if (w >= 0) {
+					subscribers[w].request(n);
+				}
+				else {
+					for (FirstNonEmptyEmittingSubscriber<T> s : subscribers) {
+						s.request(n);
+					}
+				}
+			}
+		}
+
+		@Override
+		public void cancel() {
+			if (cancelled) {
+				return;
+			}
+			cancelled = true;
+
+			int w = wip;
+			if (w >= 0) {
+				subscribers[w].cancel();
+			}
+			else {
+				for (FirstNonEmptyEmittingSubscriber<T> s : subscribers) {
+					s.cancel();
+				}
+			}
+		}
+
+		boolean tryWin(int index) {
+			if (wip == Integer.MIN_VALUE) {
+				if (WIP.compareAndSet(this, Integer.MIN_VALUE, index)) {
+
+					FirstNonEmptyEmittingSubscriber<T>[] a = subscribers;
+					int n = a.length;
+
+					for (int i = 0; i < n; i++) {
+						if (i != index) {
+							a[i].cancel();
+						}
+					}
+
+					return true;
+				}
+			}
+			return false;
+		}
+
+		void resignFromRace() {
+			COMPETING_SUBSCRIBERS.decrementAndGet(this);
+		}
+
+		boolean isLastRemainingFlux() {
+			return COMPETING_SUBSCRIBERS.get(this) == 0;
+		}
+
+	}
+
+	static final class FirstNonEmptyEmittingSubscriber<T>
+			extends Operators.DeferredSubscription implements InnerOperator<T, T> {
+
+		final RaceCoordinator<T> parent;
+
+		final CoreSubscriber<? super T> actual;
+
+		final int index;
+
+		boolean won;
+
+		FirstNonEmptyEmittingSubscriber(CoreSubscriber<? super T> actual,
+				RaceCoordinator<T> parent, int index) {
+			this.actual = actual;
+			this.parent = parent;
+			this.index = index;
+		}
+
+		@Override
+		@Nullable
+		public Object scanUnsafe(Attr key) {
+			if (key == Attr.PARENT) {
+				return s;
+			}
+			if (key == Attr.CANCELLED) {
+				return parent.cancelled;
+			}
+
+			return InnerOperator.super.scanUnsafe(key);
+		}
+
+		@Override
+		public void onSubscribe(Subscription s) {
+			set(s);
+		}
+
+		@Override
+		public CoreSubscriber<? super T> actual() {
+			return actual;
+		}
+
+		@Override
+		public void onNext(T t) {
+			if (won) {
+				actual.onNext(t);
+			}
+			else if (parent.tryWin(index)) {
+				won = true;
+				actual.onNext(t);
+			}
+		}
+
+		@Override
+		public void onError(Throwable t) {
+			if (won) {
+				actual.onError(t);
+			}
+			else if (parent.tryWin(index)) {
+				won = true;
+				actual.onError(t);
+			}
+		}
+
+		@Override
+		public void onComplete() {
+			parent.resignFromRace();
```

Comment:
use the returned valued (see previous comment) to atomically decide to propagate complete